src/server/main_service.cc

```diff
@@ -1038,6 +1038,79 @@ void Service::Function(CmdArgList args, ConnectionContext* cntx) {
   return (*cntx)->SendError(err, kSyntaxErrType);
 }
 
+vector<std::string_view> UnionResultVec(const vector<vector<string_view>>& result_vec) {
+  absl::flat_hash_set<std::string_view> uniques;
+
+  for (const auto& val : result_vec) {
+    for (const string_view& s : val) {
+      uniques.emplace(s);
+    }
+  }
+
+  vector<std::string_view> result;
+  result.reserve(uniques.size());
+  copy(uniques.begin(), uniques.end(), back_inserter(result));
+  return result;
+}
+
+void Service::PubsubChannels(string_view pattern, ConnectionContext* cntx) {
+  vector<vector<string_view>> result_set(shard_set->size());
+
+  shard_set->RunBriefInParallel([&](EngineShard* shard) {
+    result_set[shard->shard_id()] = shard->channel_slice().ListChannels(pattern);
+  });
+
+  (*cntx)->SendStringArr(UnionResultVec(result_set));
+}
+
+void Service::PubsubPatterns(ConnectionContext* cntx) {
+  vector<vector<string_view>> result_set(shard_set->size());
```

Comment:
I think what you do here is more complicated than it needs to be. each pattern is added to all shards. Hence if all you want is their count, then you can just query a single shard. say shard 0 using `shard_set->Await(0, ...);`. Moreover if you query a single shard then you do not need 
`ListPatterns`. You need to implement `PatternCount` function. i.e. do not need to fetch all the string and pass them. All you need is a single size_t. Simpler.